### PR TITLE
feat/AB#105619_EIS-allow-to-auto-select-first-option-of-resource-field-when-using-dynamic-filter 

### PR DESCRIPTION
**Pull request by unai-reliefapp**
# Description

feat: add new dataReadt event for core grid on data list is set in the grid 
feat: add new autoSelectFirstOption for dynamic filtering in resources

## Useful links

- [Link to ticket](https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/105619)

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [ ] Test A
- [ ] Test B

## Screenshots

image

https://github.com/user-attachments/assets/7ced8f64-87dd-4c33-b01f-e8f14b2db977

# Checklist:

( * == Mandatory ) 

- [X] * I have set myself as assignee of the pull request
- [X] * My code follows the style guidelines of this project
- [X] * Linting does not generate new warnings
- [X] * I have performed a self-review of my own code
- [X] * I have put the ticket for review, adding the oort-frontend team to the list of reviewers
- [X] * I have commented my code, particularly in hard-to-understand areas
- [X] * I have put JSDoc comment in all required places
- [X] * My changes generate no new warnings
- [X] * I have included screenshots describing my changes if relevant
- [X] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# More explanation
https://www.loom.com/share/05a716d61b9744faaf51fb304c21d1e5?sid=f87cf896-582a-4f76-93ae-8ceed801b145

### CODE REVIEW

**Comment by unai-reliefapp** [comment]
Hi @AntoineRelief , please let me know if this is the exact expected behavior(you can check the media attached to the PR too) before setting the ticket on review

KR, Unai

**Comment by AntoineRelief** [comment]
> Hi @AntoineRelief , please let me know if this is the exact expected behavior(you can check the media attached to the PR too) before setting the ticket on review
> 
> KR, Unai

Hey @unai-reliefapp 
That's kind of that but it's for **resource**, not **resources**

We don't need to update the  **resources** behavior

**Comment by unai-reliefapp** [comment]
> > Hi @AntoineRelief , please let me know if this is the exact expected behavior(you can check the media attached to the PR too) before setting the ticket on review
> > KR, Unai
> 
> Hey @unai-reliefapp That's kind of that but it's for **resource**, not **resources**
> 
> We don't need to update the **resources** behavior

Hi @AntoineRelief , we have the dynamic filtering in the resource question properties, but the resource query does not have the filter args set
image
And in the `getResourceRecordsById` method from resource.ts file in the frontend repo we are sending the filters to the resource query.
image


Is something missing? Is there a branch that I should use or it's just that the backend part is missing entirely?

**Comment by unai-reliefapp** [comment]
Also @AntoineRelief , moving the `autoSelectFirstOption` new property to the resource question would be enough for this ticket to be finished or im overdoing?

KR, Unai

**Comment by AntoineRelief** [comment]
hey @unai-reliefapp 

that's because you're looking at the resource.query level, but the arguments can be used at other levels, in the resource.type

in the backend, this is what we have in the resource.type, and it "should" make it work:
`records: {
      type: RecordConnectionType,
      args: {
        first: { type: GraphQLInt },
        afterCursor: { type: GraphQLID },
        filter: { type: GraphQLJSON },
        archived: { type: GraphQLBoolean },
      },`

yes, moving the property and testing the behavior should be enough

**Comment by unai-reliefapp** [comment]
> hey @unai-reliefapp
> 
> that's because you're looking at the resource.query level, but the arguments can be used at other levels, in the resource.type
> 
> in the backend, this is what we have in the resource.type, and it "should" make it work: `records: { type: RecordConnectionType, args: { first: { type: GraphQLInt }, afterCursor: { type: GraphQLID }, filter: { type: GraphQLJSON }, archived: { type: GraphQLBoolean }, },`
> 
> yes, moving the property and testing the behavior should be enough

I have bad news?
https://github.com/user-attachments/assets/ff84c4fb-78a4-4261-aa44-64688dd6679e

**Comment by unai-reliefapp** [comment]
> > hey @unai-reliefapp
> > that's because you're looking at the resource.query level, but the arguments can be used at other levels, in the resource.type
> > in the backend, this is what we have in the resource.type, and it "should" make it work: `records: { type: RecordConnectionType, args: { first: { type: GraphQLInt }, afterCursor: { type: GraphQLID }, filter: { type: GraphQLJSON }, archived: { type: GraphQLBoolean }, },`
> > yes, moving the property and testing the behavior should be enough
> 
> I have bad news? https://github.com/user-attachments/assets/ff84c4fb-78a4-4261-aa44-64688dd6679e

The query was built using the custom filter property, like this
```
[
    {
        "field": "title",
        "operator": "contains",
        "value": "Test 3"
    }
]
```

**Comment by AntoineRelief** [comment]
> > > hey @unai-reliefapp
> > > that's because you're looking at the resource.query level, but the arguments can be used at other levels, in the resource.type
> > > in the backend, this is what we have in the resource.type, and it "should" make it work: `records: { type: RecordConnectionType, args: { first: { type: GraphQLInt }, afterCursor: { type: GraphQLID }, filter: { type: GraphQLJSON }, archived: { type: GraphQLBoolean }, },`
> > > yes, moving the property and testing the behavior should be enough
> > 
> > 
> > I have bad news? https://github.com/user-attachments/assets/ff84c4fb-78a4-4261-aa44-64688dd6679e
> 
> The query was built using the custom filter property, like this
> 
> ```
> [
>     {
>         "field": "title",
>         "operator": "contains",
>         "value": "Test 3"
>     }
> ]
> ```

@unai-reliefapp 

continue the ticket "as if" it was working
in parallel, I'll have a look at why it doesn't work

File: libs/shared/src/lib/survey/components/resources.ts [hunk 1]
```diff
@@ -1,36 +1,36 @@
+import { Dialog } from '@angular/cdk/dialog';
```
Comment (unai-reliefapp) [inline libs/shared/src/lib/survey/components/resources.ts:1]: Order of imports from previous commit, no big deal

File: libs/shared/src/lib/components/search-resource-grid-modal/search-resource-grid-modal.component.ts [hunk 2]
```diff
@@ -70,10 +68,12 @@ export class ResourceGridModalComponent {
     public dialogRef: DialogRef<ResourceGridModalComponent>,
     private ref: ApplicationRef
   ) {
-    if (this.data.multiselect !== undefined)
+    if (this.data.multiselect !== undefined) {
```
Comment (AntoineRelief) [inline libs/shared/src/lib/components/search-resource-grid-modal/search-resource-grid-modal.component.ts:71]: @unai-reliefapp 
well spotted 👍

File: libs/shared/src/lib/survey/components/resource.ts [hunk 3]
```diff
@@ -417,9 +430,12 @@ export const init = (
 
             const customFilter = JSON.parse(question.customFilter);
             if (Array.isArray(customFilter)) {
-              question.filters = customFilter
-                .map((x) => updateFilter(surveyData, x))
-                .filter((x) => !isNil(x));
+              question.filters = {
```
Comment (AntoineRelief) [inline libs/shared/src/lib/survey/components/resource.ts:433]: @unai-reliefapp 

that was the issue
building the filters as an array of { field: etc }, it wouldn't work, so like that, we're making sure the filter is a composite filter, and it's compatible with the back-end

the back-end expects a composite filter ( logic field + filters field ), or a single filter ( field / value / operator )

File: libs/shared/src/lib/survey/components/resource.ts [hunk 4]
```diff
@@ -480,6 +496,13 @@ export const init = (
         getResourceRecordsById(question).subscribe(({ data }) => {
           const choices = mapQuestionChoices(data, question);
           question.contentQuestion.choices = choices;
+          if (
+            choices.length === 1 &&
+            !!question.customFilter &&
+            question.autoSelectFirstOption
+          ) {
+            question.value = question.contentQuestion.choices[0].value;
```
Comment (AntoineRelief) [inline libs/shared/src/lib/survey/components/resource.ts:504]: @unai-reliefapp 
small issue there, it was building an array but the field needs a single value

Comment (unai-reliefapp) [reply libs/shared/src/lib/survey/components/resource.ts:504]: > @unai-reliefapp small issue there, it was building an array but the field needs a single value

I thought to see a multiselect option, but there is not, you are right, thanks 👍

**Pull request merged**